src/Interfaces/FwupdManager.vala

```diff
@@ -0,0 +1,317 @@
+                    break;
+                default:
+                    break;
+            }
+        });
+
+        return release;
+    }
+
+    private string get_path (string uri) {
+        var parts = uri.split ("/");
+        string file_path = parts[parts.length - 1];
+        return Path.build_path (Path.DIR_SEPARATOR_S, Environment.get_tmp_dir (), file_path);
+    }
+
+    public async string? download_file (Device device, string uri) {
+        var path = get_path (uri);
+
+        File server_file = File.new_for_uri (uri);
+        File local_file = File.new_for_path (path);
+
+        if (FileUtils.test (path, FileTest.IS_REGULAR)) {
+            return path;
+        }
+
+        bool result;
+        try {
+            result = yield server_file.copy_async (local_file, FileCopyFlags.OVERWRITE, Priority.DEFAULT, null, (current_num_bytes, total_num_bytes) => {});
+        } catch (Error e) {
+            on_device_error (device, "Could not download file: %s".printf (e.message));
+            return null;
+        }
+
+        if (!result) {
+            on_device_error (device, "Download of %s was not succesfull".printf (uri));
+            return null;
+        }
+
+        return path;
+    }
+
+    public async bool install (Device device, string path) {
+        try {
+            // https://github.com/fwupd/fwupd/blob/c0d4c09a02a40167e9de57f82c0033bb92e24167/libfwupd/fwupd-client.c#L2045
+            var options = new GLib.HashTable<string, Variant> (str_hash, str_equal);
+            options.insert ("reason", new Variant.string ("user-action"));
+            options.insert ("filename", new Variant.string (path));
+            options.insert ("allow-older", new Variant.boolean (true));
+            options.insert ("allow-reinstall", new Variant.boolean (true));
+            options.insert ("no-history", new Variant.boolean (true));
+
+            var fd = ro_fd (path);
+            var handle = new UnixInputStream (fd, true);
+
+            yield fwupd.install (device.id, handle, options);
+        } catch (Error e) {
+            warning ("Could not connect to fwupd interface: %s", e.message);
+            on_device_error (device, device.update_error != null ? device.update_error : e.message);
+            return false;
+        }
+
+        return true;
+    }
+
+    public async Details get_details (Device device, string path) {
+        var details = new Details ();
+
+        try {
+            var fd = ro_fd (path);
+            var handle = new UnixInputStream (fd, true);
+
+            var result = yield fwupd.get_details (handle);
+            foreach (unowned GLib.HashTable<string, Variant> serialized_details in result) {
+                serialized_details.@foreach ((key, val) => {
+                    if (key == "Release") {
+                        var iter = val.iterator ().next_value ().iterator ();
+                        while (iter.next ("{sv}", out key, out val)) {
```

Comment:
I don't like that we're overwrite `key` and `val` of the outer loop in an inner loop here. It's probably fine, but the code would be clearer if you used different variables.